test/e2e/e2e.go

```diff
@@ -1357,187 +1388,216 @@ var _ = ginkgo.Describe("e2e multiple ecmp external gateway validation", func()
-			gwLo := fmt.Sprintf("%s%d", exGWRemoteIpPrefix, lastOctet)
-			go func() {
-				_, err = framework.RunKubectl(f.Namespace.Name, "exec", srcPingPodName, testContainerFlag, "--", "ping", "-c", testTimeout, gwLo)
+		for _, address := range addresses.targetIPs {
+			go func(target string) {
+				_, err := framework.RunKubectl(f.Namespace.Name, "exec", srcPodName, testContainerFlag, "--", "ping", "-c", testTimeout, target)
 				if err != nil {
-					framework.Logf("error generating a ping from the test pod %s: %v", srcPingPodName, err)
+					framework.Logf("error generating a ping from the test pod %s: %v", srcPodName, err)
 				}
-			}()
+			}(address)
+		}
+
+		err := g.Wait()
+		framework.ExpectNoError(err, "failed to reach the mock gateway(s)")
+	}, ginkgotable.Entry("IPV4", &addressesv4, "icmp"),
+		ginkgotable.Entry("IPV6", &addressesv6, "icmp6"))
+
+	// This test runs a listener on the external container, returning the host name both on tcp and udp.
+	// The src pod tries to hit the remote address until both the containers are hit.
+	ginkgotable.DescribeTable("Should validate connectivity to multiple external gateways for a UDP / TCP scenario", func(addresses *ipFamilyTestAddresses, protocol string, destPort int) {
+		if addresses.srcPodIP == "" || addresses.nodeIP == "" {
+			skipper.Skipf("Skipping as pod ip / node ip are not set pod ip %s node ip %s", addresses.srcPodIP, addresses.nodeIP)
 		}
 
-		// collect any errors and report them in a failure report
-		errs := []error{}
-		for i := 0; i < 2; i++ {
-			if err := <-icmpChan; err != nil {
-				errs = append(errs, err)
+		// annotate the test namespace with multiple gateways defined
+		annotateArgs := []string{
+			"annotate",
+			"namespace",
+			f.Namespace.Name,
+			fmt.Sprintf("k8s.ovn.org/routing-external-gws=%s,%s", addresses.gatewayIPs[0], addresses.gatewayIPs[1]),
+		}
+		framework.Logf("Annotating the external gateway test namespace to container gateways: %s, %s", addresses.gatewayIPs[0], addresses.gatewayIPs[1])
+		framework.RunKubectlOrDie(f.Namespace.Name, annotateArgs...)
+
+		expectedHostNames := make(map[string]struct{})
+		for _, c := range []string{gwContainer1, gwContainer2} {
+			res, err := runCommand("docker", "exec", c, "hostname")
+			framework.ExpectNoError(err, "failed to run hostname on", c)
+			hostname := strings.TrimSuffix(res, "\n")
+			expectedHostNames[hostname] = struct{}{}
+		}
+		framework.Logf("Expected hostnames are %v", expectedHostNames)
+
+		returnedHostNames := make(map[string]struct{})
+		success := false
+
+		// Picking only the first address, the one the udp listener is set for
+		target := addresses.targetIPs[0]
+		for i := 0; i < 20; i++ {
+			args := []string{"exec", srcPodName, testContainerFlag, "--"}
+			if protocol == "tcp" {
+				args = append(args, "bash", "-c", fmt.Sprintf("echo | nc -w 1 %s %d", target, destPort))
+			} else {
+				args = append(args, "bash", "-c", fmt.Sprintf("echo | nc -w 1 -u %s %d", target, destPort))
```

Comment:
heh, we realised some days ago that this success only if it is able to send the packet, it doesn't guarantee that the port is open, you can try against a non-existent IP
```
echo | nc -w 1 -u 5.5.5.5 1232 && echo OK
```

Reply:
but still, we check that the other end returns its hostname, so I guess it's fine

Reply:
ah, I missed that